cassandra/src/main/scala/zio/flow/internal/CassandraKeyValueStore.scala

```diff
@@ -0,0 +1,172 @@
+package zio.flow.internal
+
+import com.datastax.oss.driver.api.core.cql.{AsyncResultSet, Row, SimpleStatement}
+import com.datastax.oss.driver.api.core.{CqlIdentifier, CqlSession}
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder.literal
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder
+
+import java.io.IOException
+import java.nio.ByteBuffer
+import zio.{Chunk, Has, IO, URLayer, ZIO}
+import zio.stream.ZStream
+
+final class CassandraKeyValueStore(session: CqlSession) extends KeyValueStore {
+  import CassandraKeyValueStore._
+
+  private val keyspace =
+    session.getKeyspace
+      .orElse(null: CqlIdentifier) // scalafix:ok DisableSyntax.null
+
+  private val table =
+    CqlIdentifier.fromCql(CassandraKeyValueStore.tableName)
+
+  private val cqlSelect =
+    QueryBuilder.selectFrom(keyspace, table)
+
+  private val cqlInsert =
+    QueryBuilder.insertInto(keyspace, table)
+
+  override def put(
+    namespace: String,
+    key: Chunk[Byte],
+    value: Chunk[Byte]
+  ): IO[IOException, Boolean] = {
+    val insert = cqlInsert
+      .value(
+        namespaceColumnName,
+        literal(namespace)
+      )
+      .value(
+        keyColumnName,
+        byteBufferFrom(key)
+      )
+      .value(
+        valueColumnName,
+        byteBufferFrom(value)
+      )
+      .build
+
+    executeAsync(insert, session)
+      .mapBoth(
+        refineToIOException(s"Error putting key-value pair for <$namespace> namespace"),
+        _ => true
+      )
+  }
+
+  override def get(namespace: String, key: Chunk[Byte]): IO[IOException, Option[Chunk[Byte]]] = {
+    val query = cqlSelect
+      .column(valueColumnName)
+      .whereColumn(namespaceColumnName)
+      .isEqualTo(
+        literal(namespace)
+      )
+      .whereColumn(keyColumnName)
+      .isEqualTo(
+        byteBufferFrom(key)
+      )
+      .limit(1)
+      .build
+
+    executeAsync(query, session).flatMap { result =>
+      if (result.remaining > 0)
+        ZIO.some(
+          blobValueOf(valueColumnName, result.one)
+        )
+      else
+        ZIO.none
+    }.mapError(
+      refineToIOException(s"Error retrieving or reading value for <$namespace> namespace")
+    )
+  }
+
+  override def scanAll(namespace: String): ZStream[Any, IOException, (Chunk[Byte], Chunk[Byte])] = {
+    val query = cqlSelect
+      .column(keyColumnName)
+      .column(valueColumnName)
+      .whereColumn(namespaceColumnName)
+      .isEqualTo(
+        literal(namespace)
+      )
+      .allowFiltering
+      .build
+
+    ZStream
+      .paginateChunkM(
+        executeAsync(query, session)
+      )(_.map { result =>
+        val pairs = byteChunkPairsFrom(result)
```

Comment:
I wonder if we can even make this return a `ZStream`. Currently this `byteChunkPairsFrom` helper function converts a Java iterator into a `Chunk`. If `ZStream.paginate*` can take a `ZStream` for the paging operation, then we can stream out each row instead of using `paginateChunk*`.

@adamgfraser Is it possible?

Reply:
I'm attempting to make my own `paginateStreamM` :)

Reply:
Got it working :)